massa-versioning-worker/src/versioning.rs

```diff
@@ -0,0 +1,1080 @@
+
+    fn try_from(value: [(MipInfo, MipState); N]) -> Result<Self, Self::Error> {
+        MipStoreRaw::try_from(value)
+            .and_then(|store_raw| Ok(Self(Arc::new(RwLock::new(store_raw)))))
+    }
+}
+
+/// Store of all versioning info
+#[derive(Debug, Clone, Default)]
+pub struct MipStoreRaw(pub(crate) BTreeMap<MipInfo, MipState>);
+
+impl MipStoreRaw {
+    /// Update our store with another (usually after a bootstrap where we received another store)
+    /// Return true if update is successful, false if something is wrong on given store raw
+    fn update_with(&mut self, store_raw: &MipStoreRaw) -> bool {
+        // iter over items in given store:
+        // -> 2 cases: VersioningInfo is already in self store -> add to 'to_update' list
+        //             VersioningInfo is not in self.store -> add to 'to_add' list
+        //
+
+        let mut component_versions: HashMap<MipComponent, u32> = self
+            .0
+            .iter()
+            .map(|i| (i.0.component.clone(), i.0.component_version))
+            .collect();
+        let mut names: BTreeSet<String> = self.0.iter().map(|i| i.0.name.clone()).collect();
+        let mut to_update: BTreeMap<MipInfo, MipState> = Default::default();
+        let mut to_add: BTreeMap<MipInfo, MipState> = Default::default();
+        let mut should_merge = true;
+
+        for (v_info, v_state) in store_raw.0.iter() {
+            if !v_state.is_coherent_with(v_info) {
+                // As soon as we found one non coherent state we abort the merge
+                should_merge = false;
+                break;
+            }
+
+            if let Some(v_state_orig) = self.0.get(v_info) {
+                // Versioning info (from right) is already in self (left)
+                // Need to check if we add this to 'to_update' list
+                let v_state_id: u32 = ComponentStateTypeId::from(&v_state.inner).into();
+                let v_state_orig_id: u32 = ComponentStateTypeId::from(&v_state_orig.inner).into();
+
+                if matches!(
+                    v_state_orig.inner,
+                    ComponentState::Defined(_)
+                        | ComponentState::Started(_)
+                        | ComponentState::LockedIn(_)
+                ) {
+                    // Only accept 'higher' state
+                    // (e.g. 'started' if 'defined', 'locked in' if 'started'...)
+                    if v_state_id > v_state_orig_id {
+                        to_update.insert(v_info.clone(), v_state.clone());
+                    } else {
+                        // Trying to downgrade state' (e.g. trying to go from 'active' -> 'defined')
+                        should_merge = false;
+                        break;
+                    }
+                }
+            } else {
+                // Versioning info (from right) is not in self.0 (left)
+                // Need to check if we add this to 'to_add' list
+
+                let last_v_info_ = to_add
```

Comment:
Don't really understand the logic on this code. Why do you check only the last element of the store ? It could be possible that we want to add a version that start before another existing ?

Reply:
The only case I see is running a older version of the code and bootstrapping to a server running a higher version of the code. In that cases, we can add some new MipInfo but it will always be after existing one. I do not see any case where you can add a version that start before another existing...